follow HTTP redirects to fix login in different countries

anyc opened the pull request:
### Check if this PR fulfills these requirements:
* [x] My changes respect the [Kodi add-on development rules](https://kodi.wiki/view/Add-on_rules)
* [x] I have read the [**CONTRIBUTING**](../../master/Contributing.md) document.
* [x] I made sure there wasn't another [Pull Request](../../../pulls) opened for the same update/change
* [x] I have successfully tested my changes locally

#### Types of changes
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Feature change (non-breaking change which changes behaviour of an existing functionality)
- [ ] Improvement (non-breaking change which improves functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (non-breaking performance or readability improvements)

### Description
<!--- Motivation and a possible detailed explanation of the changes -->
<!--- Put your text below this line -->

In my country, I could not use this plugin anymore due to a similar error message as described in #1425. However, the fixes there did not help in my case. I checked the log and I saw that the the /login endpoint returned a redirect to /de-en/login in my case. In this commit, I added code to follow redirects in the make_http_call() function. An additional change is necessary in calls to httpx which is also present in the existing PR #1485.

### In case of Feature change / Breaking change:
#### Describe the current behavior
<!--- Put your text below this line -->

#### Describe the new behavior
<!--- Put your text below this line -->

### Screenshots (if appropriate):
<!--- Add some screenshots if they can be useful to show the differences between before and after the changes -->

CastagnaIT commented on resources/lib/common/ipc.py:
```diff
@@ -101,22 +101,30 @@ def make_http_call(endpoint, func_name, data=None):
+                            raise _data
+                        return _data
+                return None
+            except HTTPError as exc:
+                if exc.status == 302:
+                    url = urljoin(url, e.headers['Location'])
+                    continue
+                raise exc
```
this must not be done here
you are mixing IPC and http

anyc replied:
Hm, I do not understand what you mean here. Isn't it doing a HTTP request here that gets a 302 response that my new code handles? I only briefly checked the existing code, so I might be missing something here.

CastagnaIT replied:
this is the IPC module and dont have to manage problems on http requests
this must be implemented in different way in to http module

CastagnaIT commented on resources/lib/services/nfsession/session/base.py:
```diff
@@ -44,7 +44,7 @@ def _init_session(self):
         # (http1=False, http2=True) means that the client know that server support HTTP/2 and avoid to do negotiations,
         # prior knowledge: https://python-hyper.org/projects/hyper-h2/en/v2.3.1/negotiating-http2.html#prior-knowledge
-        self.session = httpx.Client(http1=False, http2=True)
+        self.session = httpx.Client(http1=False, http2=True, follow_redirects=True)
```
`follow_redirects` on this httpx version is already set as true

anyc replied:
But the default changes depending on the httpx version, as far as I understood. With this patch the code works in any case.

CastagnaIT replied:
you cant change follow_redirects in the old version so add this is useless

CastagnaIT commented:
> I could not use this plugin anymore due to a similar error message as described in https://github.com/CastagnaIT/plugin.video.netflix/issues/1425

what message? specify
what the reason why you need to follow the redirect if the cookies are already provided in the response?

anyc commented:
This is the part of the log:
```
2023-01-01 21:50:16.750 T:15670    INFO <general>: Loading skin file: DialogKeyboard.xml, load type: KEEP_IN_MEMORY
2023-01-01 21:50:26.976 T:15777   ERROR <general>: [plugin.video.netflix (0)] Traceback (most recent call last):
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/services/nfsession/session/access.py", line 147, in login
                                                       react_context = website.extract_json(self.get('login'), 'reactContext')
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/services/nfsession/session/http_requests.py", line 34, in get
                                                       return self._request_call(
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/utils/logging.py", line 139, in timing_wrapper
                                                       return func(*args, **kwargs)
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/services/nfsession/session/http_requests.py", line 48, in _request_call
                                                       return self._request(method, endpoint, None, **kwargs)
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/services/nfsession/session/http_requests.py", line 113, in _request
                                                       response.raise_for_status()
                                                     File "/lib32/python3.8/site-packages/httpx/_models.py", line 1508, in raise_for_status
                                                       raise HTTPStatusError(message, request=request, response=self)
                                                   httpx.HTTPStatusError: Redirect response '302 Found' for url 'https://www.netflix.com/login'
                                                   Redirect location: 'https://www.netflix.com/de-en/login'
                                                   For more information check: https://httpstatuses.com/302
                                                   
2023-01-01 21:50:26.977 T:15777   ERROR <general>: [plugin.video.netflix (0)] IPC callback raised exception: Redirect response '302 Found' for url 'https://www.netflix.com/login'
                                                   Redirect location: 'https://www.netflix.com/de-en/login'
                                                   For more information check: https://httpstatuses.com/302
2023-01-01 21:50:26.979 T:15777   ERROR <general>: [plugin.video.netflix (0)] Traceback (most recent call last):
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/services/http_server.py", line 105, in handle_request
                                                       ret_data = _call_func(func, pickle.loads(data))
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/services/http_server.py", line 161, in _call_func
                                                       return func(**data)
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/utils/logging.py", line 139, in timing_wrapper
                                                       return func(*args, **kwargs)
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/services/nfsession/session/access.py", line 147, in login
                                                       react_context = website.extract_json(self.get('login'), 'reactContext')
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/services/nfsession/session/http_requests.py", line 34, in get
                                                       return self._request_call(
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/utils/logging.py", line 139, in timing_wrapper
                                                       return func(*args, **kwargs)
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/services/nfsession/session/http_requests.py", line 48, in _request_call
                                                       return self._request(method, endpoint, None, **kwargs)
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/services/nfsession/session/http_requests.py", line 113, in _request
                                                       response.raise_for_status()
                                                     File "/lib32/python3.8/site-packages/httpx/_models.py", line 1508, in raise_for_status
                                                       raise HTTPStatusError(message, request=request, response=self)
                                                   httpx.HTTPStatusError: Redirect response '302 Found' for url 'https://www.netflix.com/login'
                                                   Redirect location: 'https://www.netflix.com/de-en/login'
                                                   For more information check: https://httpstatuses.com/302
                                                   
2023-01-01 21:50:28.196 T:15706   ERROR <general>: OpenConnection - could not opening a connection to the CEC adapter
2023-01-01 21:50:31.309 T:15774   ERROR <general>: [plugin.video.netflix (1)] Traceback (most recent call last):
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/run_addon.py", line 29, in wrapper
                                                       func(*args, **kwargs)
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/run_addon.py", line 265, in run
                                                       success = route(pathitems)
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/run_addon.py", line 88, in lazy_login_wrapper
                                                       if G.REQUEST_PARAMS.get('ignore_login') or _check_valid_credentials():
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/run_addon.py", line 73, in _check_valid_credentials
                                                       if not login():
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/utils/api_requests.py", line 36, in login
                                                       if common.make_call('login', credentials):
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/utils/logging.py", line 139, in timing_wrapper
                                                       return func(*args, **kwargs)
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/common/ipc.py", line 94, in make_call
                                                       return make_http_call(endpoint, func_name, data)
                                                     File "/usr/share/kodi/addons/plugin.video.netflix/resources/lib/common/ipc.py", line 117, in make_http_call
                                                       _data = pickle.loads(received_data)
                                                   TypeError: __init__() missing 2 required keyword-only arguments: 'request' and 'response'
                                                   
2023-01-01 21:50:31.315 T:15670    INFO <general>: Loading skin file: DialogConfirm.xml, load type: KEEP_IN_MEMORY
2023-01-01 21:50:33.548 T:15774    INFO <general>: CPythonInvoker(6, /usr/share/kodi/addons/plugin.video.netflix/addon.py): script successfully run
2023-01-01 21:50:33.548 T:15773   ERROR <general>: GetDirectory - Error getting plugin://plugin.video.netflix/
2023-01-01 21:50:33.582 T:15670   ERROR <general>: CGUIMediaWindow::GetDirectory(plugin://plugin.video.netflix/) failed
2023-01-01 21:50:48.203 T:15706   ERROR <general>: OpenConnection - could not opening a connection to the CEC adapter
2023-01-01 21:51:08.209 T:15706   ERROR <general>: OpenConnection - could not opening a connection to the CEC adapter
```
It was my first guess that the redirection might be the issue. I added the code and the plugin worked again. Before, I also tried to remove the plugin, cleaned the userdata and reinstall it again, but the error stayed the same.

anyc commented:
Btw., if no one else has this issue and the patch looks odd to you, I don't mind if you close the PR. I can apply this patch only in my build system.

CastagnaIT commented:
> Btw., if no one else has this issue and the patch looks odd to you, I don't mind if you close the PR. I can apply this patch only in my build system.

not odd, not right implemented
i tried login but to me for now works
may be a change that may become effective to all accounts in the coming days or weeks

anyc commented:
Maybe the `/login` to `/de-en/login` redirection happens because I'm located in Germany but use English as default language on most devices.

CastagnaIT commented:
> Maybe the `/login` to `/de-en/login` redirection happens because I'm located in Germany but use English as default language on most devices.

i will try test by switching default language to see if i catch same problem
however this redirect is not new there for some years

CastagnaIT commented:
i tried but redirect on login do not happens in to my country
but the request should follow redirect automatically without doing it manually
maybe we have only to avoid to raise the status

can you try login by changing:
https://github.com/CastagnaIT/plugin.video.netflix/blob/master/resources/lib/services/nfsession/session/http_requests.py#L113
to:
```py
if response.status_code != 302:
      response.raise_for_status()
```

CastagnaIT commented:
no user has complained about this problem last months,
in addition to being a completely incorrect implementation,
the redirect is already enabled on the httpx module
deeper checks should be made to know why you need force route the request,
but since you are strangely the only singular case,
and also i cannot replicate the problem in any way i am closing PR
